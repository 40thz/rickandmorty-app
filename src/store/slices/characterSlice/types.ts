import { DataStatus } from '../types';

export enum CHARACTER_STATUS {
  ALIVE = 'Alive',
  DEAD = 'Dead',
  UNKNOWN = 'unknown',
}

export enum CHARACTER_GENDER {
  FEMALE = 'Female',
  MALE = 'Male',
  GENDERLESS = 'Genderless',
  UNKNOWN = 'unknown',
}

export enum DATA_TYPE {
  TABLE = 'Table',
  GRID = 'Grid',
}

type Origin = {
  name: string;
  url: string;
};

export type Character = {
  id: number;
  name: string;
  status: CHARACTER_STATUS.ALIVE | CHARACTER_STATUS.DEAD | CHARACTER_STATUS.UNKNOWN;
  species: string;
  type: string;
  gender:
    | CHARACTER_GENDER.FEMALE
    | CHARACTER_GENDER.GENDERLESS
    | CHARACTER_GENDER.GENDERLESS
    | CHARACTER_GENDER.UNKNOWN;
  origin: Origin;
  location: Origin;
  image: string;
  episode: string[];
  url: string;
  created: string;
};

type CharacterOptions = {
  name: string;
  species: string;
  type: string;
  gender: string;
  status: string;
};

export type CharacterState = {
  type: DATA_TYPE;
  options: CharacterOptions;
  data: Character[] | null;
  status: DataStatus;
};
