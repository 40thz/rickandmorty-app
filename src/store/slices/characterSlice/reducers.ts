import { createAsyncThunk } from '@reduxjs/toolkit';
import { characterService } from '@/services/character.service';
import { CharacterState } from './types';

export const characterFind = createAsyncThunk('character/find', async (_, { getState }) => {
  const {
    character: { options },
  } = getState() as { character: CharacterState };

  try {
    const query = `?${new URLSearchParams(options as Record<string, string>).toString()}`;

    return await characterService.find(query);
  } catch (error) {
    return Promise.reject(error);
  }
});
