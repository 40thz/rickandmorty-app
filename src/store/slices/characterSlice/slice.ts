import { createSlice } from '@reduxjs/toolkit';
import { DATA_MODE } from '../types';
import { characterFind } from './reducers';
import { CharacterState } from './types';

export const initialState: CharacterState = {
  mode: DATA_MODE.TABLE,
  options: {
    name: '',
    species: '',
    type: '',
    gender: '',
    status: '',
  },
  data: [],
  status: {
    isLoading: true,
    error: null,
  },
};

const characterSlice = createSlice({
  name: 'character',
  initialState,
  reducers: {
    setOptions(state, action) {
      state.options = { ...state.options, ...action.payload };
    },
    setDataMode(state, { payload }) {
      state.mode = payload;
    },
  },

  extraReducers: (builder) => {
    builder.addCase(characterFind.pending, (state) => {
      state.status.isLoading = true;
    });

    builder.addCase(characterFind.fulfilled, (state, { payload: { results } }) => {
      state.data = results;
      state.status.isLoading = false;
      state.status.error = null;
    });

    builder.addCase(characterFind.rejected, (state, action) => {
      state.data = [];
      state.status.isLoading = true;
      state.status.error = action.error.message;
    });
  },
});

export const { setOptions, setDataMode } = characterSlice.actions;

export const characherReducer = characterSlice.reducer;
