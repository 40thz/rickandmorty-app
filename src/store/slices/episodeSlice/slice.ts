import { createSlice } from '@reduxjs/toolkit';
import { DATA_MODE } from '../types';
import { episodeFind } from './reducers';
import { EpisodeState } from './types';

export const initialState: EpisodeState = {
  mode: DATA_MODE.TABLE,
  options: {
    name: '',
    episode: '',
  },
  data: [],
  status: {
    isLoading: true,
    error: null,
  },
};

const episodeSlice = createSlice({
  name: 'episode',
  initialState,
  reducers: {
    setOptions(state, action) {
      state.options = { ...state.options, ...action.payload };
    },
    setDataMode(state, { payload }) {
      state.mode = payload;
    },
  },

  extraReducers: (builder) => {
    builder.addCase(episodeFind.pending, (state) => {
      state.status.isLoading = true;
    });

    builder.addCase(episodeFind.fulfilled, (state, { payload: { results } }) => {
      state.data = results;
      state.status.isLoading = false;
      state.status.error = null;
    });

    builder.addCase(episodeFind.rejected, (state, action) => {
      state.data = [];
      state.status.isLoading = true;
      state.status.error = action.error.message;
    });
  },
});

export const { setOptions, setDataMode } = episodeSlice.actions;

export const episodeReducer = episodeSlice.reducer;
