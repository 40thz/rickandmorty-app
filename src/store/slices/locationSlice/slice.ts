import { createSlice } from '@reduxjs/toolkit';
import { DATA_MODE } from '../types';
import { locationFind } from './reducers';
import { LocationState } from './types';

const initialState: LocationState = {
  mode: DATA_MODE.TABLE,
  options: {
    name: '',
    type: '',
    dimension: '',
  },
  data: [],
  status: {
    isLoading: true,
    error: null,
  },
};

const locationSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {
    setOptions(state, action) {
      state.options = { ...state.options, ...action.payload };
    },
    setDataMode(state, { payload }) {
      state.mode = payload;
    },
  },

  extraReducers: (builder) => {
    builder.addCase(locationFind.pending, (state) => {
      state.status.isLoading = true;
    });

    builder.addCase(locationFind.fulfilled, (state, { payload: { results } }) => {
      state.data = results;
      state.status.isLoading = false;
      state.status.error = null;
    });

    builder.addCase(locationFind.rejected, (state, action) => {
      state.data = [];
      state.status.isLoading = true;
      state.status.error = action.error.message;
    });
  },
});

export const { setOptions, setDataMode } = locationSlice.actions;

export const locationReducer = locationSlice.reducer;
