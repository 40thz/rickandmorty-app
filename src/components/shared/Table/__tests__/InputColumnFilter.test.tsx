import * as reduxHooks from 'react-redux';
import * as actions from '@@/store/slices/characterSlice/slice';
import { fireEvent, render, screen } from '@testing-library/react';
import { InputColumnFilter } from '../';

jest.mock('react-redux');

describe('InputColumnFilter component', () => {
  const mockedDispath = jest.spyOn(reduxHooks, 'useDispatch');

  it('Should be render', () => {
    const mockedSetOptions = jest.spyOn(actions, 'setOptions');
    jest.spyOn(reduxHooks, 'useDispatch');

    const component = render(<InputColumnFilter setOptions={mockedSetOptions as never} prop="name" value="123" />);

    expect(component).toMatchSnapshot();
  });

  it('Should dispath action setOptions', () => {
    const dispath = jest.fn();
    mockedDispath.mockReturnValue(dispath);
    const mockedSetOptions = jest.spyOn(actions, 'setOptions');

    render(<InputColumnFilter setOptions={mockedSetOptions as never} prop="name" value="123" type="checkbox" />);

    fireEvent.click(screen.getByRole('checkbox'));

    expect(dispath).toHaveBeenCalledTimes(1);
    expect(mockedSetOptions).toHaveBeenCalledWith({ name: '123' });
  });
});
