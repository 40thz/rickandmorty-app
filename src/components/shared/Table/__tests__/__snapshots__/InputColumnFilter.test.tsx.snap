// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputColumnFilter component Should be render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label>
        <input
          class="text-white/55 rounded-[6px] block w py-1 px-2 bg-gray w-full"
          id="default-input"
          type="text"
          value="123"
        />
      </label>
    </div>
  </body>,
  "container": <div>
    <label>
      <input
        class="text-white/55 rounded-[6px] block w py-1 px-2 bg-gray w-full"
        id="default-input"
        type="text"
        value="123"
      />
    </label>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
